pinot-core/src/main/java/org/apache/pinot/core/operator/blocks/InstanceResponseUtils.java

```diff
@@ -0,0 +1,226 @@
+import org.apache.pinot.core.operator.blocks.results.ExceptionResultsBlock;
+import org.apache.pinot.core.operator.blocks.results.MetadataResultsBlock;
+import org.apache.pinot.core.query.aggregation.function.AggregationFunction;
+import org.apache.pinot.core.query.aggregation.function.DistinctAggregationFunction;
+import org.apache.pinot.core.query.distinct.DistinctTable;
+import org.apache.pinot.core.query.request.context.QueryContext;
+import org.apache.pinot.core.query.request.context.utils.QueryContextUtils;
+
+
+public final class InstanceResponseUtils {
+  private InstanceResponseUtils() {
+    // do not instantiate.
+  }
+
+  public static InstanceResponseBlock getEmptyResponseBlock(DataSchema explainResultSchema) {
+    return new InstanceResponseBlock(new MetadataResultsBlock(explainResultSchema), null);
+  }
+
+  public static InstanceResponseBlock getExceptionBlock(ProcessingException exception) {
+    return new InstanceResponseBlock(new ExceptionResultsBlock(exception), null);
+  }
+
+  public static DataTable toDataTable(InstanceResponseBlock instanceResponseBlock) {
+    DataTable dataTable;
+    try {
+      if (instanceResponseBlock.getBaseResultsBlock() == null) {
+        dataTable = DataTableBuilderUtils.buildEmptyDataTable();
+      } else {
+         dataTable = instanceResponseBlock.getBaseResultsBlock().getDataTable(instanceResponseBlock.getQueryContext());
+      }
+      if (MapUtils.isNotEmpty(instanceResponseBlock.getExceptionMap())) {
+        for (Map.Entry<Integer, String> e : instanceResponseBlock.getExceptionMap().entrySet()) {
+          dataTable.addException(e.getKey(), e.getValue());
+        }
+      }
+      if (MapUtils.isNotEmpty(instanceResponseBlock.getInstanceResponseMetadata())) {
+        dataTable.getMetadata().putAll(instanceResponseBlock.getInstanceResponseMetadata());
+      }
+      return dataTable;
+    } catch (ProcessingException pe) {
+      return toDataTable(getExceptionBlock(pe));
+    } catch (Exception e) {
+      return toDataTable(getExceptionBlock(QueryException.UNKNOWN_ERROR));
+    }
+  }
+
+  public static Collection<Object[]> toRows(InstanceResponseBlock instanceResponseBlock) {
+    try {
+      return instanceResponseBlock.getBaseResultsBlock().getRows(instanceResponseBlock.getQueryContext());
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while building data table", e);
+    }
+  }
+
+  public static DataTable getDataSchema(InstanceResponseBlock instanceResponseBlock) {
+    try {
+      return instanceResponseBlock.getBaseResultsBlock().getDataTable(instanceResponseBlock.getQueryContext());
+    } catch (Exception e) {
+      throw new RuntimeException("Caught exception while building data table", e);
+    }
+  }
+
+  /**
+   * Builds an empty data table based on the broker request.
+   */
+  public static InstanceResponseBlock buildEmptyResultBlock(QueryContext queryContext)
```

Comment:
I am guessing all these below are moved as is from DataTableBuilderUtils (just so I can skip reviewing them )?

Reply:
yes